### PR TITLE
Private claims

### PR DESCRIPTION
**Pull request by djoey123**
Hi,

Thanks for looking at the pull request :)
This is my first time working on a Minecraft plugin / Java,
I would love to get some feedback :)

I added the ability to set a claim to private / public.
I added it in the config file to enable / disable, default it is disabled

### CODE REVIEW

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java [hunk 1]
```diff
@@ -460,6 +460,29 @@ public boolean isBannedFromClaim(@NotNull OnlineUser user, @NotNull Claim claim,
         return false;
     }
 
+    // Check if a claim is private for user
+    @ApiStatus.Internal
+    public boolean CheckIfClaimIsPrivate(@NotNull OnlineUser user, @NotNull Claim claim, @NotNull HuskClaims plugin) {
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java:465]: Let's clarify this method name a bit - canAccessPrivateClaim perhaps?

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java [hunk 2]
```diff
@@ -460,6 +460,29 @@ public boolean isBannedFromClaim(@NotNull OnlineUser user, @NotNull Claim claim,
...
+        if (!claim.isPrivateClaim()) {
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java:466]: A small optimisation here for servers with this off - let's return true if private claims are disabled in the plugin config to begin with :)

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorldSerializer.java [hunk 3]
```diff
@@ -109,7 +109,8 @@ protected static Claim getUpgradableClaim(@NotNull JsonObject co, @NotNull Gson
         childrenArray.forEach(c -> children.add(getUpgradableClaim(c.getAsJsonObject(), gson)));
         final Set<OperationType> defaultFlags = gson.fromJson(co.getAsJsonArray("default_flags"), GsonProvider.OPERATION_TYPE_SET_TOKEN);
         final boolean inheritParent = co.get("inherit_parent").getAsBoolean();
-        return new Claim(owner, region, trustedUsers, trustedGroups, trustedTags, Maps.newConcurrentMap(), children, inheritParent, defaultFlags);
+        final boolean privateClaim = co.get("private_claim").getAsBoolean();
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/claim/ClaimWorldSerializer.java:112]: This logic here is for upgrading claims from older data versions - hence this isn't needed as this is a new feature; let's get rid of this and just use "false" for the constructor call :)

File: common/src/main/java/net/william278/huskclaims/config/Settings.java [hunk 4]
```diff
@@ -281,6 +281,10 @@ public static class BanSettings {
                     "Also requires the MANAGE_BANS privilege (by default, restricted to those with 'manage' trust)"})
             private boolean enabled = false;
 
+            @Comment({"Whether to let users set their claim claims to private (prevent others from entering) using /claimprivate",
+                    "Also requires the MANAGE_BANS privilege (by default, restricted to those with 'manage' trust)"})
+            private boolean private_claims = false;
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/config/Settings.java:286]: Let's name this variable in camel case - privateClaims

File: common/src/main/resources/locales/en-gb.yml [hunk 5]
```diff
@@ -219,4 +219,6 @@ locales:
   transferclaim_command_description: 'Transfer ownership of the claim you''re standing in.'
   unclaimall_command_description: 'Delete all of your claims.'
   transferpet_command_description: 'Transfer ownership of a tamed animal.'
-  claimban_command_description: 'Ban a player from a claim.'
\ No newline at end of file
+  claimban_command_description: 'Ban a player from a claim.'
+  claim_private_enabled: 'Claim is set to private.'
+  claim_private_disabled: 'Claim is set to public.'
```
Comment (WiIIiam278) [inline common/src/main/resources/locales/en-gb.yml:224]: Let's do two things to these two new locale keys (in every locale file):

- move them in the locale file to below the keys for claim bans
- add color formatting in the standard color (the bluey/green one I use) by wrapping the text with `[brackets](#color) `

The new command description locale key can stay put.

File: common/src/main/java/net/william278/huskclaims/claim/Claim.java [hunk 6]
```diff
@@ -140,6 +140,17 @@ public class Claim implements Highlightable {
     @SerializedName("inherit_parent")
     private boolean inheritParent;
 
+    /**
+     * If this is a child claim, whether to inherit member trust levels from the parent.
+     * <p>
+     * If set to false, this child claim will be restricted.
+     */
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/claim/Claim.java:147]: Don't forget to update this JavaDoc comment :)

**Review by WiIIiam278 — CHANGES_REQUESTED** [review]
Hey, thanks for contributing! :)

> This is my first time working on a Minecraft plugin / Java,

Well, you've done an amazing job if that's the case! Just a few notes for consistency really. Excellent work.

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java (hunk 1 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java]: Indeed, much better :)

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java (hunk 2 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java:470]: Added then check :)

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorldSerializer.java (hunk 3 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/claim/ClaimWorldSerializer.java]: 👍

File: common/src/main/java/net/william278/huskclaims/config/Settings.java (hunk 4 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/config/Settings.java]: 👍

File: common/src/main/resources/locales/en-gb.yml (hunk 5 above)
Comment (djoey123) [reply common/src/main/resources/locales/en-gb.yml]: Moved and added the green color :)

File: common/src/main/java/net/william278/huskclaims/claim/Claim.java (hunk 6 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/claim/Claim.java]: Updated the JavaDoc :)

**Comment by djoey123** [comment]
> Hey, thanks for contributing! :)
> 
> > This is my first time working on a Minecraft plugin / Java,
> 
> Well, you've done an amazing job if that's the case! Just a few notes for consistency really. Excellent work.
> 
> There is only one thing to consider - when the user makes the home private we'll need to consider logic for kicking untrusted users from the claim. There's a method I made to do this for banning a user - perhaps you'd like to give this a go? If not no worries as this can be done by me later.

Thanks :D

I updated the pull request with the feedback.
I will take a look at the kicking part when the users are still in the claim.

File: common/src/main/java/net/william278/huskclaims/command/ClaimPrivateCommand.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import net.william278.huskclaims.trust.TrustLevel;
+import net.william278.huskclaims.user.OnlineUser;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+
+public class ClaimPrivateCommand extends InClaimCommand {
+
+    protected ClaimPrivateCommand(@NotNull HuskClaims plugin) {
+        super(List.of("claimprivate"), "", TrustLevel.Privilege.MANAGE_BANS, plugin);
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/command/ClaimPrivateCommand.java:34]: Just thinking about this privilege here - I think we should add a separate privilege called `MAKE_PRIVATE` for a bit more customizability there (default to being usable by only the claim owner).

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java [hunk 8]
```diff
@@ -460,6 +460,33 @@ public boolean isBannedFromClaim(@NotNull OnlineUser user, @NotNull Claim claim,
...
+
+        if (isIgnoringClaims(user, plugin)) {
+            return false;
+        }
+        
+        if (claim.getTrustLevel(user, plugin).isPresent()) {
+            return false;
+        }
+
+        return true;
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java:487]: Just a bit of a cleanup here to simplify things.
```suggestion
        final Optional<UUID> owner = claim.getOwner();
        if (owner.isPresent() && owner.get().equals(user.getUuid()) || isIgnoringClaims(user, plugin)) {
            return false;
        }
        return claim.getTrustLevel(user, plugin).isEmpty();
```
I've also realised this method name is now inconsistent with the boolean value we're returning - so perhaps lets go with `isBlockedFromPrivateClaim`

**Comment by WiIIiam278** [comment]
> I will take a look at the kicking part when the users are still in the claim.

Sure, just let me know when you're done if happy.

File: common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java (hunk 8 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/claim/ClaimWorld.java]: Changed it :)

File: common/src/main/java/net/william278/huskclaims/command/ClaimPrivateCommand.java (hunk 7 above)
Comment (djoey123) [reply common/src/main/java/net/william278/huskclaims/command/ClaimPrivateCommand.java]: Added the privilege

**Comment by djoey123** [comment]
I added the kicking out of the claim :)
It seems the private status is not saved on server restart.
Did I do something wrong? or am I missing something?

**Comment by djoey123** [comment]
Fixed the saving problem :)

plugin.getDatabase().updateClaimWorld(world);

was missing

**Comment by djoey123** [comment]
Also added the login check like the claim ban :)

**Review by WiIIiam278 — APPROVED** [review]

**Comment by WiIIiam278** [comment]
Great work!

**Pull request merged**